A2rchi/app/app.py

```diff
@@ -0,0 +1,427 @@
+            except FileNotFoundError:
+                sources = dict()
+
+            # get the closest source to the document
+            source = None
+            if len(result['source_documents']) > 0:
+                source_hash = result['source_documents'][0].metadata['source']
+                if '/' in source_hash and '.' in source_hash:
+                    source = source_hash.split('/')[-1].split('.')[0]
+
+            # if the score is low enough, include the source as a link, otherwise give just the answer
+            embedding_name = self.config["utils"]["embeddings"]["EMBEDDING_NAME"]
+            similarity_score_reference = self.config["utils"]["embeddings"]["EMBEDDING_CLASS_MAP"][embedding_name]["similarity_score_reference"]
+            if score < similarity_score_reference and source in sources.keys(): 
+                output = "<p>" + result["answer"] + "</p>" + "\n\n<br /><br /><p><a href= " + sources[source] + ">Click here to read more</a></p>"
+            else:
+                output = "<p>" + result["answer"] + "</p>"
+
+            ChatWrapper.update_or_add_discussion(self.data_path, "conversations_test.json", discussion_id, history)
+
+        except Exception as e:
+            raise e
+        finally:
+            self.lock.release()
+        return output, discussion_id
+
+
+# Flask app setup
+app = Flask("test-app")
+app.secret_key = os.environ.get("SECRET_KEY") or os.urandom(24)
+
+app.config["TEMPLATE_FOLDER"] = "A2rchi/app/templates"
+app.config["STATIC_FOLDER"] = "A2rchi/app/static"
+
+# create the chat from the wrapper
+# chat = ChatWrapper()
+
+# determine whether we're configuring login or guest access
+global_config = Config_Loader().config["global"]
+USE_LOGIN = global_config["USE_LOGIN"]
+ALLOW_GUEST_LOGIN = global_config["ALLOW_GUEST_LOGIN"]
+GOOGLE_LOGIN = global_config["GOOGLE_LOGIN"]
+MIT_LOGIN = global_config["MIT_LOGIN"]
+if USE_LOGIN:
+    # get set of valid users
+    VALID_USER_EMAILS = global_config["USER_EMAILS"] if USE_LOGIN else []
+
+    # User session management setup
+    # https://flask-login.readthedocs.io/en/latest
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    # Flask-Login helper to retrieve a user from our db
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User.get(user_id)
+    
+    # Naive database setup
+    try:
+        init_db_command()
+    except sqlite3.OperationalError:
+        # Assume it's already been created
+        pass
+
+    # OAuth 2 client setup
+    google_client, mit_client = None, None
+    if GOOGLE_LOGIN:
+        google_client = WebApplicationClient(GOOGLE_CLIENT_ID)
+    if MIT_LOGIN:
+        mit_client = WebApplicationClient(MIT_CLIENT_ID)
+
+    def get_google_provider_cfg():
+        return requests.get(GOOGLE_DISCOVERY_URL).json()
+
+    def get_mit_provider_cfg():
+        return requests.get(MIT_DISCOVERY_URL).json()
+
+# enable CORS:
+CORS(app)
```

Comment:
Eventually will want to move this back into a wrapper class. But for now I think it looks like it works and we can do this later.

Reply:
Moved back into the wrapper